### PR TITLE
Tr bdf2

### PR DESCRIPTION
**Pull request by atb1995**
Poorly named branch that implements various timestepping schemes in Gusto: 
TR-BDF2 (two stage implicit)
Adams-Moulton Family up to AM4 (multistep implicit)
Leapfrog  and Adams Bashforth (up to AB5) (multistep explicit)

Adams-Bashforth required a shorter timestep for the integration test (they have tighter stability constraints than Forward Euler), and Leapfrog requires CG advection rather than DG. I looked at DG0 upwind leapfrog and think that the computational mode is unconditionally unstable.

I also did some convergence analysis on TR-BDF2 and other timestepping schemes for the linear advection equation, they converge to the correct order.

### CODE REVIEW

File: gusto/time_discretisation.py [hunk 1]
```diff
@@ -1084,3 +1083,440 @@ def apply(self, x_out, *x_in):
...
+
+    def apply(self, x_out, *x_in):
+        """
+        Apply the time discretisation to advance one whole time step.
+
+        Args:
+            x_out (:class:`Function`): the output field to be computed.
+            x_in (:class:`Function`): the input field(s).
+        """
+        if self.initial_timesteps < self.nlevels-2:
```
Comment (jshipton) [inline gusto/time_discretisation.py:1350]: I think that this means that your `self.solver0` is never used...

Comment (atb1995) [reply gusto/time_discretisation.py]: Thanks Jemma! Addressed this

**Comment by atb1995** [comment]
convergence_cfl_0 2_degree_2

File: gusto/time_discretisation.py [hunk 2]
```diff
@@ -1084,3 +1090,463 @@ def apply(self, x_out, *x_in):
...
+
+    def apply(self, x_out, *x_in):
+        """
+        Apply the time discretisation to advance one whole time step.
+
+        Args:
+            x_out (:class:`Function`): the output field to be computed.
+            x_in (:class:`Function`): the input field(s).
+        """
+        if self.initial_timesteps < self.nlevels-1:
```
Comment (tommbendall) [inline gusto/time_discretisation.py:1544]: We should make an issue about how to deal with these time steppers when checkpointing

File: gusto/time_discretisation.py [hunk 3]
```diff
@@ -1084,3 +1090,463 @@ def apply(self, x_out, *x_in):
...
+        else:
+            solver = self.solver
+        self.xnm2.assign(x_in[0])
+        self.xnm1.assign(x_in[1])
+        self.x1.assign(x_in[2])
+        solver.solve()
+        x_out.assign(self.x_out)
+
+
+class AM2(MultilevelTimeDiscretisation):
```
Comment (tommbendall) [inline gusto/time_discretisation.py:1463]: AdamsMoulton2 ?

File: gusto/time_discretisation.py [hunk 4]
```diff
@@ -1084,3 +1090,463 @@ def apply(self, x_out, *x_in):
...
+        else:
+            solver = self.solver
+
+        self.xnm1.assign(x_in[0])
+        self.x1.assign(x_in[1])
+        solver.solve()
+        x_out.assign(self.x_out)
+
+
+class AB2(MultilevelTimeDiscretisation):
```
Comment (tommbendall) [inline gusto/time_discretisation.py:1303]: Could we rename these to make it a bit clearer? i.e. `AdamsBashforth2`?

File: gusto/time_discretisation.py [hunk 5]
```diff
@@ -1084,3 +1090,463 @@ def apply(self, x_out, *x_in):
...
+            solver = self.solver0
+        else:
+            solver = self.solver
+        self.xnm1.assign(x_in[0])
+        self.x1.assign(x_in[1])
+        solver.solve()
+        x_out.assign(self.x_out)
+
+
+class AB3(MultilevelTimeDiscretisation):
```
Comment (tommbendall) [inline gusto/time_discretisation.py:1380]: Is there a general way of writing AB2 and AB3? Is it best to implement them as different classes or could they be the same class with an optional argument to indicate the number of stages?

**Review by tommbendall — COMMENTED** [review]
I think this is looking good, it's well-documented and tested. My main questions are about the names of the new classes and whether any of them can be combined?

File: gusto/time_discretisation.py (hunk 2 above)
Comment (atb1995) [reply gusto/time_discretisation.py:1574]: Yep agreed! I have made one

File: gusto/time_discretisation.py (hunk 4 above)
Comment (atb1995) [reply gusto/time_discretisation.py]: Done

File: gusto/time_discretisation.py (hunk 5 above)
Comment (atb1995) [reply gusto/time_discretisation.py]: I've created two more general time discretisations AdamsMoulton and AdamsBashforth that take in order as an argument, and run the corresponding method. We now have up to AB5 and AM4

File: gusto/time_discretisation.py [hunk 6]
```diff
@@ -1084,3 +1090,495 @@ def apply(self, x_out, *x_in):
...
+        Raises:
+            ValueError: if order is not provided, or is in incorrect range.
+        """
+
+        if (order > 5 or order < 1):
+            raise ValueError("Adams-Bashforth of order greater than 5 not implemented")
+        if isinstance(options, (EmbeddedDGOptions, RecoveryOptions)):
+            raise NotImplementedError("Only SUPG advection options have been implemented for this time discretisation")
+        if not solver_parameters:
+            # theta method leads to asymmetric matrix, per lhs function below,
```
Comment (tommbendall) [inline gusto/time_discretisation.py:1335]: Were these solver options just taken from the theta method? Maybe we should update the comment?

File: gusto/time_discretisation.py [hunk 7]
```diff
@@ -1084,3 +1090,495 @@ def apply(self, x_out, *x_in):
...
+
+        Raises:
+            ValueError: if order is not provided, or is in incorrect range.
+        """
+        if (order > 4 or order < 1):
+            raise ValueError("Adams-Moulton of order greater than 5 not implemented")
+        if isinstance(options, (EmbeddedDGOptions, RecoveryOptions)):
+            raise NotImplementedError("Only SUPG advection options have been implemented for this time discretisation")
+        if not solver_parameters:
+            # theta method leads to asymmetric matrix, per lhs function below,
```
Comment (tommbendall) [inline gusto/time_discretisation.py:1468]: Same comment here about theta method

File: gusto/time_discretisation.py [hunk 8]
```diff
@@ -1084,3 +1090,495 @@ def apply(self, x_out, *x_in):
...
+            ValueError: if gamma is not provided.
+        """
+        # TODO: would this be better as a non-optional argument? Or should the
+        # check be on the provided value?
+        if gamma is None:
+            raise ValueError("please provide a value for gamma between 0 and 1")
+        if isinstance(options, (EmbeddedDGOptions, RecoveryOptions)):
+            raise NotImplementedError("Only SUPG advection options have been implemented for this time discretisation")
+        if not solver_parameters:
+            # theta method leads to asymmetric matrix, per lhs function below,
```
Comment (tommbendall) [inline gusto/time_discretisation.py:1131]: I spotted this below. Presumably you have taken the default solver parameters from the theta method? At the very least we should update the reference to theta method here, but maybe we should think about what our default solver options should be

File: gusto/time_discretisation.py [hunk 9]
```diff
@@ -1084,3 +1090,495 @@ def apply(self, x_out, *x_in):
...
+                                            map_if_true=drop,
+                                            map_if_false=replace_subject(self.x[n], self.idx))
+            rtemp = rtemp.label_map(lambda t: t.has_label(time_derivative),
+                                    map_if_false=lambda t: -self.dt*self.br[n]*t)
+            r += rtemp
+        return r.form
+
+    @property
+    def solver0(self):
+        """Set up the problem and the solver."""
```
Comment (tommbendall) [inline gusto/time_discretisation.py:1552]: Could you add a comment to clarify that this is for the first steps? Ideally we should also do this for all the other lhs0/rhs0/solver0 routines

File: gusto/time_discretisation.py [hunk 10]
```diff
@@ -1084,3 +1090,495 @@ def apply(self, x_out, *x_in):
...
+            self.br = [(1.0/2.0)]
+        elif (self.order == 2):
+            self.bl = (5.0/12.0)
+            self.br = [-(1.0/12.0), (8.0/12.0)]
+        elif (self.order == 3):
+            self.bl = (9.0/24.0)
+            self.br = [(1.0/24.0), -(5.0/24.0), (19.0/24.0)]
+        elif (self.order == 4):
+            self.bl = (251.0/720.0)
+            self.br = [-(19.0/720.0), (106.0/720.0), -(254.0/720.0), (646.0/720.0)]
```
Comment (tommbendall) [inline gusto/time_discretisation.py:1497]: I will presume that all of these coefficients are correct!

File: gusto/time_discretisation.py [hunk 11]
```diff
@@ -1084,3 +1090,495 @@ def apply(self, x_out, *x_in):
...
+        if (self.order == 1):
+            self.b = [1.0]
+        elif (self.order == 2):
+            self.b = [-(1.0/2.0), (3.0/2.0)]
+        elif (self.order == 3):
+            self.b = [(5.0)/(12.0), -(16.0)/(12.0), (23.0)/(12.0)]
+        elif (self.order == 4):
+            self.b = [-(9.0)/(24.0), (37.0)/(24.0), -(59.0)/(24.0), (55.0)/(24.0)]
+        elif (self.order == 5):
+            self.b = [(251.0)/(720.0), -(1274.0)/(720.0), (2616.0)/(720.0), -(2774.0)/(720.0), (2901.0)/(720.0)]
```
Comment (tommbendall) [inline gusto/time_discretisation.py:1362]: I will presume that all of these coefficients are correct! Did you manage to test the convergence somewhere?

File: gusto/time_discretisation.py [hunk 12]
```diff
@@ -1084,3 +1090,495 @@ def apply(self, x_out, *x_in):
...
         solver.solve()
         x_out.assign(self.x_out)
+
+
+class TR_BDF2(TimeDiscretisation):
+    """
+    Implements the two stage implicit TR-BDF2 timestepping method, with a trapeziodal stage follwed
+    by a second order backwards difference stage.
+
+    The TR_BDF2 timestepping method for operator F is written as
```
[the remark this answers is not in the document]
Comment (atb1995) [reply gusto/time_discretisation.py:1100]: Yes, trapezoidal (TR) and second order backwards difference (BDF2). Hopefully shown in new description

File: gusto/time_discretisation.py [hunk 13]
```diff
@@ -1084,3 +1090,495 @@ def apply(self, x_out, *x_in):
...
+    y^(n+g) = y^n + dt*g*F[y^n] + dt*g*F[y^(n+g)] (TR stage)
+    y^(n+1) = 1/(g(2-g))*y^(n+g) - (1-g)**2/(g(2-g))*y^(n) + (1-g)/(2-g)*dt*F[y^(n+1)] (BDF2 stage)
+    for parameter g (gamma).
```
[the remark this answers is not in the document]
Comment (atb1995) [reply gusto/time_discretisation.py:1103]: Done!

File: gusto/time_discretisation.py [hunk 14]
```diff
@@ -1084,3 +1090,495 @@ def apply(self, x_out, *x_in):
...
+
+        r = (1.0/(self.gamma*(2.0-self.gamma)))*xnpg - ((1.0-self.gamma)**2/(self.gamma*(2.0-self.gamma))) * xn
+
+        return r.form
+
+    @cached_property
+    def solver_tr(self):
+        """Set up the problem and the solver."""
+        # setup solver using lhs and rhs defined in derived class
+        problem = NonlinearVariationalProblem(self.lhs-self.rhs, self.x_out, bcs=self.bcs)
```
[the remark this answers is not in the document]
Comment (atb1995) [reply gusto/time_discretisation.py:1201]: Yep, have done so

File: gusto/time_discretisation.py (hunk 6 above)
Comment (atb1995) [reply gusto/time_discretisation.py:1335]: Removed the comment! I guess this is just some default solver settings, I hadn't given it thought of what it should be..

File: gusto/time_discretisation.py (hunk 8 above)
Comment (atb1995) [reply gusto/time_discretisation.py:1131]: I agree.. I just went with the theta method ones as I hadn't given it much thought. What options are available?

File: gusto/time_discretisation.py (hunk 9 above)
Comment (atb1995) [reply gusto/time_discretisation.py:1552]: I have done this for all schemes where I have added this initial solver and rhs/lhs

File: gusto/time_discretisation.py (hunk 11 above)
Comment (atb1995) [reply gusto/time_discretisation.py:1362]: For this and previous comment: I believe so, I will do some convergence tests for all schemes to check and post here

**Review by tommbendall — APPROVED** [review]
Thanks Alex, all looking good now!

**Pull request merged**